unittests/runtime/Concurrent.cpp

```diff
@@ -42,3 +44,48 @@ TEST(ConcurrentReadableArrayTest, SingleThreaded) {
   add(1000000);
   check();
 }
+
+TEST(ConcurrentReadableArrayTest, MultiThreaded) {
+  const int writerCount = 16;
+  const int readerCount = 8;
+  const int insertCount = 100000;
+
+  struct Value {
+    int threadNumber;
+    int x;
+  };
+  ConcurrentReadableArray<Value> array;
+
+  auto writer = [&](int threadNumber) {
+    for (int i = 0; i < insertCount; i++)
+      array.push_back({ threadNumber, i });
+  };
+
+  auto reader = [&] {
+    int maxByThread[writerCount];
+    bool done = false;
+    while (!done) {
+      for (int i = 0; i < writerCount; i++)
+        maxByThread[i] = -1;
+      for (auto element : array.snapshot()) {
```

Comment:
I wish I saw an easy way to do more snapshot operations.  Each reader takes a snapshot and then does quite a bit of work to verify the contents.  This suffices to verify that reading the snapshot is independent of the ongoing write operations, but snapshotting itself seems like the bigger risk.  Since the readers are doing a fair bit of work here, I wouldn't be surprised if the writers all finished before the readers had done more than one or two iterations, so you might have no more than 16-24 snapshot operations.

Not a blocker; this will get run on a lot of CI passes so any issues should pop up eventually.